# htseq2multx

Wrapper to fastq-multx to have it mimmic barcode_splitter and thus easily integrate with Princeton's HTSEQ web interface

## Description

The Princeton HTSEQ software is a web access to data generated by the Core Sequencing Facility in the Lewis Sigler Institute for Integrative Genomics.  It provides functionality for upstream analysis and data download and transfer.

Part of the supported functionality is the demultiplexing of pooled genomic samples.  The initial implementation of this functionality utilized an in-house python tool called [barcode_splitter](https://bitbucket.org/princeton_genomics/barcode_splitter/) written in the interpreted scripting language called Python.  The performance of barcode_splitter appears to be I/O bound, and as sample capacity grows and costs fall, has proven to not scale well with this growth.

fastq-multx is a demultiplexing tool written in C and performs several orders of magnitude faster.  Since HTSEQ is highly integrated with barcode_splitter, creating a wrapper to make fastq-multx look and act like barcode_splitter requires much less effort than altering HTSEQ to use fastq-multx.

